internal/driver/controllerserver_helper.go

```diff
@@ -0,0 +1,328 @@
+func (s *ControllerServer) canAttach(ctx context.Context, instance *linodego.Instance) (canAttach bool, err error) {
+	log := logger.GetLogger(ctx)
+	log.V(4).Info("Checking if volume can be attached", "instance_id", instance.ID)
+
+	// Get the maximum number of volume attachments allowed for the instance
+	limit, err := s.maxAllowedVolumeAttachments(ctx, instance)
+	if err != nil {
+		return false, err
+	}
+
+	// List the volumes currently attached to the instance
+	volumes, err := s.client.ListInstanceVolumes(ctx, instance.ID, nil)
+	if err != nil {
+		return false, errInternal("list instance volumes: %v", err)
+	}
+
+	// Return true if the number of attached volumes is less than the limit
+	return len(volumes) < limit, nil
+}
+
+// maxAllowedVolumeAttachments calculates the maximum number of volumes that can be attached to a Linode instance,
+// taking into account the instance's memory and currently attached disks.
+func (s *ControllerServer) maxAllowedVolumeAttachments(ctx context.Context, instance *linodego.Instance) (int, error) {
+	log := logger.GetLogger(ctx)
+	log.V(4).Info("Calculating max volume attachments")
+
+	// Check if the instance or its specs are nil
+	if instance == nil || instance.Specs == nil {
+		return 0, errNilInstance
+	}
+
+	// Retrieve the list of disks currently attached to the instance
+	disks, err := s.client.ListInstanceDisks(ctx, instance.ID, nil)
+	if err != nil {
+		return 0, errInternal("list instance disks: %v", err)
+	}
+
+	// Convert the reported memory from MB to bytes
+	memBytes := uint(instance.Specs.Memory) << 20
+	return maxVolumeAttachments(memBytes) - len(disks), nil
+}
+
+// getContentSourceVolume retrieves information about the Linode volume to clone from.
+// It returns a LinodeVolumeKey if a valid source volume is found, or an error if the source is invalid.
+func (cs *ControllerServer) getContentSourceVolume(ctx context.Context, contentSource *csi.VolumeContentSource) (*linodevolumes.LinodeVolumeKey, error) {
+	log := logger.GetLogger(ctx)
+	log.V(4).Info("Attempting to get content source volume")
+
+	if contentSource == nil {
+		return nil, nil // Return nil if no content source is provided
+	}
+
+	// Check if the content source type is a volume
+	if _, ok := contentSource.GetType().(*csi.VolumeContentSource_Volume); !ok {
+		return nil, errUnsupportedVolumeContentSource
+	}
+
+	sourceVolume := contentSource.GetVolume()
+	if sourceVolume == nil {
+		return nil, errNoSourceVolume // Return error if no source volume is specified
+	}
+
+	// Parse the volume ID from the content source
+	volumeInfo, err := linodevolumes.ParseLinodeVolumeKey(sourceVolume.GetVolumeId())
+	if err != nil || volumeInfo == nil {
+		return nil, errInternal("parse volume info from content source: %v", err)
+	}
+
+	// Retrieve the volume data using the parsed volume ID
+	volumeData, err := cs.client.GetVolume(ctx, volumeInfo.VolumeID)
+	if err != nil || volumeData == nil {
```

Comment:
I think we should have separate checks. If err is nil and volumeData is nil as well, the error message will not be of much help
```
get volume <nil>: <nil>"
```